editor/store/reducer.js

```diff
@@ -114,12 +114,12 @@ function getFlattenedBlocks( blocks ) {
 export const editor = flow( [
 	combineReducers,
 
-	// Track undo history, starting at editor initialization.
-	partialRight( withHistory, { resetTypes: [ 'SETUP_NEW_POST', 'SETUP_EDITOR' ] } ),
+	// Track undo history, starting at editor initialization of blocks. Assumes
+	// one of either block resetting, new post setup occurs in initialization.
+	partialRight( withHistory, { resetTypes: [ 'SETUP_NEW_POST', 'RESET_BLOCKS' ] } ),
 
-	// Track whether changes exist, resetting at each post save. Relies on
-	// editor initialization firing post reset as an effect.
-	partialRight( withChangeDetection, { resetTypes: [ 'SETUP_NEW_POST', 'RESET_POST' ] } ),
+	// Track whether changes exist, resetting at initialization and each save.
+	partialRight( withChangeDetection, { resetTypes: [ 'RESET_POST', 'SETUP_NEW_POST', 'RESET_BLOCKS' ] } ),
```

Comment:
RESET_BLOCKS is also triggered in the text editor and it should not reset dirty detection in that case

Reply:
>RESET_BLOCKS is also triggered in the text editor and it should not reset dirty detection in that case

Sigh... maybe we need a dedicated "editor ready" action in addition to the `SETUP_EDITOR` one 😅

Reply:
Should switching editors affect history? It's kind of a destructive action that, as a user, I wouldn't expect from just switching editor modes.

Reply:
> Should switching editors affect history? It's kind of a destructive action that, as a user, I wouldn't expect from just switching editor modes.

I'd look into preserving history across modes. I've experienced content loss firsthand through something like that (Calypso editor): some Heisenbug in Visual makes content disappear, toggle to Code, notice it's empty, come back to Visual, find no edit history to recover the content.